### PR TITLE
Parallel EVM Tests

### PR DESCRIPTION
**Pull request by gakonst**
Co-authored with @brockelmore, converts unit & fuzz tests to be run in parallel.

This achieves a significant speedup, especially for fuzz tests.

Things in this PR:
1. Instead of passing EVM into the runner, pass `EvmOpts` (refactor EvmOpts to evm-adapters)
2. Instantiate an EVM on each test (via `new_sputnik_evm`, this tightly couples us to SputnikEVM a bit, but should be fine, we weren't using evmodin either way)
3. Use the `SharedBackend` (instead of `ForkMemoryBackend`), which is cheap to clone, but allows sharing state across forked tests, deduplicating work as much as possible
4. Instantiating an EVM per test lets us make our functions `&self` instead of `&mut self`, which makes parallelization with `rayon` very easy: we just replace `.iter()` with `par_iter()`.

Supersedes https://github.com/gakonst/foundry/pull/435
Also contains the fixes from https://github.com/gakonst/foundry/pull/430.

* [x] Benchmark
     * @brockelmore tested `spells-mainnet` from 2m37 go to 1m43 (~30% improv)
     * [`guni-lev`](https://github.com/hexonaut/guni-lev/) from to 33s to 27.7s (~20% improvement)
     * [`drai`](https://github.com/mds1/drai) from ~18s to ~9s (2x improv)
     * **[`solmate`](https://github.com/Rari-Capital/solmate) from 2m48 to 47s (3.5x)**
* [x] Figure out the todo-item in `evm.debug_calls()` in the CLI's `run.rs` (https://github.com/gakonst/foundry/pull/445)

### CODE REVIEW

File: cli/src/cmd/run.rs [hunk 1]
```diff
@@ -123,7 +137,9 @@ impl Cmd for RunArgs {
                 })
                 .collect();
 
-            let calls = evm.debug_calls();
+            // let calls = evm.debug_calls();
+            // TODO: How should we instantiate this now that we don't have an EVM?
```
Comment (brockelmore) [inline cli/src/cmd/run.rs:141]: we will need to return the `DebugArena` in the `TestResult` now like we do for traces. This is probably how it always should have been done anyway

File: cli/src/cmd/run.rs [hunk 2]
```diff
@@ -149,14 +165,14 @@ impl Cmd for RunArgs {
                     if evm_opts.verbosity > 4 || !result.success {
                         // print setup calls as well
                         traces.iter().for_each(|trace| {
-                            trace.pretty_print(0, &known_contracts, &mut ident, runner.evm, "");
+                            trace.pretty_print(0, &known_contracts, &mut ident, &vm(), "");
```
Comment (brockelmore) [inline cli/src/cmd/run.rs:168]: `pretty_print` likely needs refactoring anyway, and since we always call `update_identified` inside a test, we no longer need the vm

Comment (brockelmore) [reply cli/src/cmd/run.rs:168]: likely this should be done in a separate PR. `pretty_print` should just build up a string that can be printed in one shot

**Review by brockelmore — APPROVED** [review]
LGTM gg

**Comment by sambacha** [comment]
what are your thoughts for this item @gakonst 

- Test for edge cases with **SharedBackend**

**Comment by gakonst** [comment]
Manual testing for now, not sure if we can properly integration test & benchmark yet.

File: forge/src/multi_runner.rs [hunk 3]
```diff
@@ -116,46 +103,57 @@ impl MultiContractRunnerBuilder {
...
-            .map(|(name, (abi, address, logs))| {
-                let result = self.run_tests(name, abi, *address, logs, filter, &init_state)?;
+            .map(|(name, (abi, deploy_code))| {
+                // unavoidable duplication here?
+                let result = match backend {
+                    BackendKind::Simple(ref backend) => {
+                        self.run_tests(name, abi, backend, deploy_code.clone(), filter)?
+                    }
+                    BackendKind::Shared(ref backend) => {
+                        self.run_tests(name, abi, backend, deploy_code.clone(), filter)?
```
Comment (mattsse) [inline forge/src/multi_runner.rs:154]: cloning is a bit expensive, 
perhaps we can at a later stage either consume on `fn test` or remove the bytecode or wrap it in an arc

Comment (gakonst) [reply forge/src/multi_runner.rs:154]: `ethers::types::Bytes` is a wrapper around `bytes::Bytes`, which is cheap to clone https://docs.rs/bytes/latest/bytes/struct.Bytes.html

**Review by mattsse — APPROVED** [review]
lgtm,
sputnik's borrowing behavior makes it a bit more verbose than it should be

**Comment by gakonst** [comment]
@brockelmore  said he'll take over the conflict & merge, else I'll wrap it in the morning
